File: 824.goat-latin.cpp
/*
 * @lc app=leetcode id=824 lang=cpp
 *
 * [824] Goat Latin
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  string toGoatLatin(string S) {
    int start = 0;
    int len = S.length();
    std::string ret;
    int index = 1;
    for (int i = 0; i < len; i++) {
      const char c = S[i];
      if (isspace(c)) {
          ret += toGoatLatinWord(S.substr(start, i - start), index++) + " ";
          start = i + 1;
      } else if (i == len - 1) {
          ret += toGoatLatinWord(S.substr(start, i - start + 1), index++);
      }
    }
    return ret;
  }

 private:
  bool isVowel(const char c) {
    return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u' ||
           c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U';
  }
  string toGoatLatinWord(const string& word, const int index) {
    if (word.length() == 0) {
      return word;
    }
    string ret;
    if (isVowel(word[0])) {
        ret += word + "m" + string(index + 1, 'a');
    } else {
        ret += word.substr(1, word.length()) + word[0] + "m" + string(index + 1, 'a');
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  ASSERT_EQ(s.toGoatLatin("I speak Goat Latin"),
            "Imaa peaksmaaa oatGmaaaa atinLmaaaaa");
  ASSERT_EQ(s.toGoatLatin("The quick brown fox jumped over the lazy dog"),
            "heTmaa uickqmaaa rownbmaaaa oxfmaaaaa umpedjmaaaaaa overmaaaaaaa "
            "hetmaaaaaaaa azylmaaaaaaaaa ogdmaaaaaaaaaa");
  return 0;
}

File: 824.goat-latin.rs
/*
 * @lc app=leetcode id=824 lang=rust
 *
 * [824] Goat Latin
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn to_goat_latin(sentence: String) -> String {
        let vowels = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'];
        let mut v = Vec::new();
        for (n, word) in sentence.split_ascii_whitespace().enumerate() {
            let trans_word = match word.starts_with(&vowels[..]) {
                true => word.to_string(),
                false => word[1..].to_string() + &word[..1],
            } + "ma";
            v.push(trans_word + &"a".repeat(n + 1))
        }
        v.join(" ")
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::to_goat_latin("I speak Goat Latin".to_string()),
        "Imaa peaksmaaa oatGmaaaa atinLmaaaaa".to_string()
    );
}

#[test]
fn test2() {
    assert_eq!(Solution::to_goat_latin("The quick brown fox jumped over the lazy dog".to_string()), "heTmaa uickqmaaa rownbmaaaa oxfmaaaaa umpedjmaaaaaa overmaaaaaaa hetmaaaaaaaa azylmaaaaaaaaa ogdmaaaaaaaaaa".to_string());
}
